elastic/beats
# Update controllermanager metrics

## What does this PR do?

This PR updates the metrics that Kubernetes `controllermamager` metricset collects.
Also the OOTB dashboard is updated.

## Why is it important?



`controllermamager ` metricset currently collects metrics from kubernetes controllermanager prometheus endpoint that have been deprecated in latest versions of kubernetes. Currently we support k8s versions starting from 1.21.x.
As a result values from those fields are always nil and some of the dashboard visualisations show no values.

Those deprecated prometheus fields  are 
```
http_request_duration_microseconds
http_request_size_bytes          
http_response_size_bytes                
http_requests_total 
```

`http_request_duration_microseconds` can instead be collected from `rest_client_request_duration_seconds`
For the rest of them, there are no more, related metrics.

Also metrics like 
`process_max_fds`, `workqueue_work_duration_seconds` are interesting and were added.

As part of this PR the following elasticsearch fields where dropped (their values were always nil in k8s versions we support)

- `kubernetes.controllermanager.http.request.duration.us.percentile.*`
- `kubernetes.controllermanager.http.request.duration.us.sum`
- `kubernetes.controllermanager.http.request.duration.us.count`
- `kubernetes.controllermanager.http.request.size.bytes.percentile.*`
- `kubernetes.controllermanager.http.request.size.bytes.sum`
- `kubernetes.controllermanager.http.request.size.bytes.count`
- `kubernetes.controllermanager.http.responsesize.bytes.percentile.*`
- `kubernetes.controllermanager.http.response.size.bytes.sum`
- `kubernetes.controllermanager.http.response.size.bytes.count`
- `kubernetes.controllermanager.http.request.count`
- `kubernetes.controllermanager.handler`

and the following fields added:

- `kubernetes.controllermanager.url`
- `kubernetes.controllermanager.verb`
- `kubernetes.controllermanager.fds.max.count`
- `kubernetes.controllermanager.client.request.duration.sec.sum`
- `kubernetes.controllermanager.client.request.duration.sec.count`
- `kubernetes.controllermanager.client.request.duration.sec.bucket.*`
- `kubernetes.controllermanager.workqueue.work.duration.sec.sum`
- `kubernetes.controllermanager.workqueue.work.duration.sec.count`
- `kubernetes.controllermanager.workqueue.work.duration.sec.bucket.*`



## Checklist




## How to test this PR locally



Deploy Metricbeat from this PR branch with controllermanager metricset enabled. 
```
- module: kubernetes
   metricsets:
       - controllermanager
   hosts: ["https://0.0.0.0:10257"]
   bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
   ssl.verification_mode: "none"
   period: 10s
```

Prefer Kind or Minikube in order for metricbeat to have access to controllermanager endpoint.
Check for the new fields and setup the new dashboards 
`./metricbeat setup --dashboards -E setup.dashboards.directory=./module/kubernetes/_meta/kibana/  -E setup.kibana.host=http://elastic:changeme@localhost:5601  -E ssl.verification_mode=none -e  -d "*"`

## Related issues

- Closes https://github.com/elastic/beats/issues/32007

## Use cases



## Screenshots

New dashboard

<img width="1743" alt="controller1" src="https://user-images.githubusercontent.com/26270880/175327940-5e6ceb37-c20e-475e-9174-e3553e9e6a5f.png">

<img width="1752" alt="controller2" src="https://user-images.githubusercontent.com/26270880/175327969-91fd81b9-ba32-4b6f-9d18-2060cac3f6af.png">
<img width="1756" alt="controller3" src="https://user-images.githubusercontent.com/26270880/175327999-403edd35-db42-4d16-9836-a17be90e9d98.png">




## Logs
